### PR TITLE
[VectorCombine] Fold "(or (zext (bitcast X)), (shl (zext (bitcast Y)), C))" -> "(bitcast (concat X, Y))" MOVMSK bool mask style patterns

### LINKED ISSUE #111431: [AVX-512] Identify opportunities for `kunpckdq`
```zig
export fn foo(a: @Vector(32, u8), b: @Vector(32, u8), c: @Vector(32, u8)) u64 {
    const b1: u32 = @bitCast(a == b);
    const b2: u32 = @bitCast(a == c);
    return (@as(u64, b1) << 32) | b2;
}
```

```llvm
define dso_local i64 @foo(<32 x i8> %0, <32 x i8> %1, <32 x i8> %2) local_unnamed_addr {
Entry:
  %3 = icmp eq <32 x i8> %0, %1
  %4 = bitcast <32 x i1> %3 to i32
  %5 = icmp eq <32 x i8> %0, %2
  %6 = bitcast <32 x i1> %5 to i32
  %7 = zext i32 %4 to i64
  %8 = shl nuw i64 %7, 32
  %9 = zext i32 %6 to i64
  %10 = or disjoint i64 %8, %9
  ret i64 %10
}
```

This compiles like so: ([Godbolt link](https://zig.godbolt.org/z/daeo4GWhr))

```asm
foo:
        vpcmpeqb        k0, ymm0, ymm1
        kmovd   ecx, k0
        vpcmpeqb        k0, ymm0, ymm2
        kmovd   eax, k0
        shl     rcx, 32
        or      rax, rcx
        vzeroupper
        ret
```

I would think it should be:

```asm
foo:
        vpcmpeqb        k0, ymm0, ymm1
        vpcmpeqb        k1, ymm0, ymm2
        kunpckdq        k0, k0, k1
        kmovd   rax, k0
        vzeroupper
        ret
```

### PR DESCRIPTION
**Pull request by RKSimon**
Mask/Bool vectors are often bitcast to/from scalar integers, in particular when concatenating mask results, often this is due to the difficulties of working with vector of bools on C/C++. On x86 this typically involves the MOVMSK/KMOV instructions.

To concatenate bool masks, these are typically cast to scalars, which are then zero-extended, shifted and OR'd together.

This patch attempts to match these scalar concatenation patterns and convert them to vector shuffles instead. This in turn often assists with further vector combines, depending on the cost model.

Fixes #111431

### CODE REVIEW

File: llvm/test/Transforms/PhaseOrdering/X86/concat-boolmasks.ll [hunk 1]
```diff
@@ -250,3 +291,5 @@ define i64 @PR111431(<32 x i8> %a0, <32 x i8> %a1, <32 x i8> %a2) {
   %or = or disjoint i64 %shl, %z02
   ret i64 %or
 }
+;; NOTE: These prefixes are unused and the list is autogenerated. Do not add tests below this line:
+; CHECK: {{.*}}
```
Comment (alexey-bataev) [inline llvm/test/Transforms/PhaseOrdering/X86/concat-boolmasks.ll:295]: Drop these lines

**Review by alexey-bataev — APPROVED** [review]
LG

**Comment by vitalybuka** [comment]
Looks like from this patch https://lab.llvm.org/buildbot/#/builders/55/builds/4146

**Comment by RKSimon** [comment]
Thanks @vitalybuka - I've now reproduced this and am working on a fix.

**Review by davemgreen — COMMENTED** [review]
Is it worth having some tests that are not in PhaseOrdering too? In case the phase ordering test changes and we don't realise we have lost the test coverage.

**Pull request merged**